stock_return_request/models/stock_return_request.py

```diff
@@ -0,0 +1,517 @@
+                self.return_from_location = warehouse.lot_stock_id.id
+
+    @api.model
+    def _default_warehouse_id(self):
+        warehouse = self.env['stock.warehouse'].search([
+            ('company_id', '=', self.env.user.company_id.id),
+        ], limit=1)
+        return warehouse
+
+    def _compute_show_to_refund(self):
+        for one in self:
+            if 'to_refund' not in self.env['stock.move']._fields:
+                continue
+            one.show_to_refund = True
+
+    def _create_picking(self, picking, picking_moves):
+        """Return Pickings for the given moves"""
+        picking_type_id = (
+            picking.picking_type_id.return_picking_type_id.id or
+            picking.picking_type_id.id)
+        new_picking = picking.copy({
+            'move_lines': [(6, 0, picking_moves.ids)],
+            'move_line_ids': [
+                (6, 0, picking_moves.mapped('move_line_ids').ids)],
+            'picking_type_id': picking_type_id,
+            'state': 'draft',
+            'origin': _("Return of %s") % picking.name,
+            'location_id': self.return_from_location.id,
+            'location_dest_id': self.return_to_location.id,
+            'stock_return_request_id': self.id,
+        })
+        new_picking.move_lines = picking_moves
+        new_picking.message_post_with_view(
+            'mail.message_origin_link',
+            values={'self': new_picking, 'origin': picking},
+            subtype_id=self.env.ref('mail.mt_note').id
+        )
+        return new_picking
+
+    def _prepare_move_default_values(self, line, qty, move):
+        """Extend this method to add values to return move"""
+        vals = {
+            'product_id': line.product_id.id,
+            'product_uom_qty': qty,
+            'product_uom': line.product_uom_id.id,
+            'state': 'draft',
+            'location_id': line.request_id.return_from_location.id,
+            'location_dest_id': line.request_id.return_to_location.id,
+            'origin_returned_move_id': move.id,
+            'procure_method': 'make_to_stock',
+        }
+        if self.show_to_refund:
+            vals['to_refund'] = line.request_id.to_refund
+        return vals
+
+    def _prepare_move_line_values(self, line, return_move, qty):
+        """Extend to add values to the move lines with lots"""
+        return {
+            'product_id': line.product_id.id,
+            'product_uom_id': line.product_uom_id.id,
+            'lot_id': line.lot_id.id,
+            'location_id': return_move.location_id.id,
+            'location_dest_id': return_move.location_dest_id.id,
+            'qty_done': qty,
+        }
+
+    def action_confirm(self):
+        """Wrapper for multi. Avoid recompute as it delays the pickings
+           creation"""
+        with self.env.norecompute():
+            for one in self:
+                one._action_confirm()
+        self.recompute()
+
+    def _action_confirm(self):
```

Comment:
Maybe split this into smaller functions.

Reply:
Some parts already are. I'll keep it in my for future revisions.